api/rest/restapi.go

```diff
@@ -362,60 +371,207 @@ func (api *API) graphHandler(w http.ResponseWriter, r *http.Request) {
 	sendResponse(w, err, graph)
 }
 
+func (api *API) consumeLocalAdd(args map[string]string, outObj *types.NodeSerial) error {
+	//TODO: when ipfs add starts supporting formats other than
+	// v0 (v1.cbor, v1.protobuf) we'll need to update this
+	b := types.BlockWithFormat{
+		Data:   outObj.Data,
+		Format: "",
+	}
+	args["cid"] = outObj.Cid // root node stored on last call
+	var hash string
+	err := api.rpcClient.Call("",
+		"Cluster",
+		"IPFSBlockPut",
+		b,
+		&hash)
+	if outObj.Cid != hash {
+		logger.Warningf("mismatch. node cid: %s\nrpc cid: %s", outObj.Cid, hash)
+	}
+	return err
+}
+
+func (api *API) finishLocalAdd(args map[string]string) error {
+	// TODO: pin the root node from here, rpc.Call(pin, args["cid"])
+	return nil
+}
+
+func (api *API) consumeShardAdd(args map[string]string, outObj *types.NodeSerial) error {
+	var shardID string
+	shardID, ok := args["id"]
+	if !ok {
+		shardID = randStringRunes(45)
+		args["id"] = shardID
+	}
+	nodeS := types.ShardNodeSerial{
+		Cid:  outObj.Cid,
+		Data: outObj.Data,
+		Size: outObj.Size,
+		ID:   shardID,
+	}
+	return api.rpcClient.Call("",
+		"Cluster",
+		"ShardAddNode",
+		nodeS,
+		&struct{}{})
+}
+
+func (api *API) finishShardAdd(args map[string]string) error {
+	shardID, ok := args["id"]
+	if !ok {
+		return errors.New("bad state: shardID passed incorrectly")
+	}
+	return api.rpcClient.Call("", "Cluster", "ShardFinalize", shardID,
+		&struct{}{})
+}
+
+func (api *API) consumeImport(ctx context.Context,
+	outChan <-chan *types.NodeSerial,
+	printChan <-chan *types.AddedOutput,
+	errChan <-chan error,
+	w http.ResponseWriter,
+	consume func(map[string]string, *types.NodeSerial) error,
+	finish func(map[string]string) error) error {
+	var err error
+	openChs := 3
+	enc := json.NewEncoder(w)
+	w.Header().Add("Content-Type", "application/json")
+	w.WriteHeader(200)
+	//	flusher, _ := w.(http.Flusher)
+	toPrint := make([]*types.AddedOutput, 0)
+	args := make(map[string]string)
+
+	for {
+		if openChs == 0 {
+			break
+		}
+
+		// Consume signals from importer.  Errors resulting from
+		// consuming the importers' signals cause error objects to be
+		// streamed in the response body.  The client reads these
+		// errors and terminates the session.
+		// Codes: 1 = error, 0 = non-error response, 2 = successful
+		// termination
+		// TODO: Unlike earlier approaches this is actually systematic
+		// and handles all possible errors.  This approach is still
+		// deferring the investigation into what error handling
+		// approaches give the best UX however.
+		select {
+		// Ensure we terminate reading from importer after cancellation
+		// but do not block
+		case <-ctx.Done():
+			err = errors.New("context timeout terminated add")
+			return enc.Encode(types.Error{Code: 1, Message: err.Error()})
```

Comment:
Because I built this with the goal of eventually streaming the response concurrent with reading the request I didn't want to assume that the header could be written at the time the error surfaced which is why errors are streamed in the body.

Maybe I am overcomplicating but from my reading of golang's http modules the client needs to know when to close the response body so I built a little application protocol over HTTP where the client knows to close on receipt of a JSON object encoded like this with an error code (1 or 2).  

If we decide against trying to get concurrent responding and request reading in the near future I can just change this to set an error response header